src/sc/sidechaintypes.cpp

```diff
@@ -1,139 +1,329 @@
 
-const libzendoomc::ScFieldElement& CompressedFieldElement::GetFieldElement() const
+////////////////////////////// Custom Config types //////////////////////////////
+bool FieldElementCertificateFieldConfig::IsValid() const
 {
-    InitFieldElement();
-    return scFieldElement;
+    if(nBits <=0 || nBits > SC_FIELD_SIZE*8)
+        return false;
+    else
+        return true;
 }
 
-#ifdef BITCOIN_TX
-bool CompressedFieldElement::IsValid() const { return true; }
-#else
-bool CompressedFieldElement::IsValid() const
+FieldElementCertificateFieldConfig::FieldElementCertificateFieldConfig(int32_t nBitsIn): CustomCertificateFieldConfig(), nBits(nBitsIn) {}
+
+int32_t FieldElementCertificateFieldConfig::getBitSize() const
+{
+    return nBits;
+}
+
+//----------------------------------------------------------------------------------
+bool BitVectorCertificateFieldConfig::IsValid() const
 {
-    InitFieldElement();
-    if (scFieldElement.IsNull())
+    bool isBitVectorSizeValid = (bitVectorSizeBits >= 0) && (bitVectorSizeBits <= MAX_BIT_VECTOR_SIZE_BITS);
+    if(!isBitVectorSizeValid)
         return false;
 
-    return libzendoomc::IsValidScFieldElement(scFieldElement);
-};
-#endif
+    bool isMaxCompressedSizeValid = (maxCompressedSizeBytes >= 0) && (maxCompressedSizeBytes <= MAX_COMPRESSED_SIZE_BYTES);
+    if(!isMaxCompressedSizeValid)
+        return false;
+
+    return true;
+}
 
-bool CompressedFieldElement::checkCfg(const CustomFieldConfig& cfg) const
+BitVectorCertificateFieldConfig::BitVectorCertificateFieldConfig(int32_t bitVectorSizeBits, int32_t maxCompressedSizeBytes): CustomCertificateFieldConfig(), bitVectorSizeBits(bitVectorSizeBits), maxCompressedSizeBytes(maxCompressedSizeBytes)
+{}
+
+
+////////////////////////////// Custom Field types //////////////////////////////
+
+
+//----------------------------------------------------------------------------------------
+FieldElementCertificateField::FieldElementCertificateField(const std::vector<unsigned char>& rawBytes)
+    :CustomCertificateField(rawBytes), pReferenceCfg{nullptr} {}
+
+FieldElementCertificateField::FieldElementCertificateField(const FieldElementCertificateField& rhs)
+    :CustomCertificateField{}, pReferenceCfg{nullptr}
 {
+    *this = rhs;
+}
+
+FieldElementCertificateField& FieldElementCertificateField::operator=(const FieldElementCertificateField& rhs)
+{
+    *const_cast<std::vector<unsigned char>*>(&vRawData) = rhs.vRawData;
+    if (rhs.pReferenceCfg != nullptr)
+        this->pReferenceCfg = new FieldElementCertificateFieldConfig(*rhs.pReferenceCfg);
+    else
+        this->pReferenceCfg = nullptr;
+    return *this;
+}
+
+
+bool FieldElementCertificateField::IsValid(const FieldElementCertificateFieldConfig& cfg) const
+{
+	return !this->GetFieldElement(cfg).IsNull();
+}
+
+const CFieldElement& FieldElementCertificateField::GetFieldElement(const FieldElementCertificateFieldConfig& cfg) const
```

Comment:
I can't get why do we need to pass `config`? why not to create FE (the only check is max size check) and then do alll the config checks inside `FieldElementCertificateField::IsValid`?

Reply:
Idea is to make sure that you'll get a Field compliant with the config you have already validated it against.